### PR TITLE
SPARK-5841 [CORE] [HOTFIX 2] Memory leak in DiskBlockManager

### PR DESCRIPTION
**Pull request by nishkamravi2**
Continue to see IllegalStateException in YARN cluster mode. Adding a simple workaround for now.

### CODE REVIEW

**Comment by nishkamravi2** [comment]
@srowen @JoshRosen @mattwhelan

**Comment by nishkamravi2** [comment]
From a recent run: 

15/02/18 22:10:49 WARN util.ShutdownHookManager: ShutdownHook '$anon$3' failed, java.lang.IllegalStateException: Shutdown in progress
java.lang.IllegalStateException: Shutdown in progress
	at java.lang.ApplicationShutdownHooks.remove(ApplicationShutdownHooks.java:82)
	at java.lang.Runtime.removeShutdownHook(Runtime.java:239)
	at org.apache.spark.storage.DiskBlockManager.stop(DiskBlockManager.scala:151)
	at org.apache.spark.storage.BlockManager.stop(BlockManager.scala:1208)
	at org.apache.spark.SparkEnv.stop(SparkEnv.scala:90)
	at org.apache.spark.SparkContext.stop(SparkContext.scala:1376)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$3.run(ApplicationMaster.scala:107)
	at org.apache.hadoop.util.ShutdownHookManager$1.run(ShutdownHookManager.java:54)

**Comment by srowen** [comment]
No, this was already fixed in https://github.com/apache/spark/commit/49c19fdbad57f0609bbcc9278f9eaa8115a73604 I don't think you have that commit yet here. The problem is not do with it being `null`; I don't think it can be and we shouldn't just let the exception occur.

**Comment by nishkamravi2** [comment]
I have that commit. The problem continues to manifest itself. This is a workaround, not a fix at the source. I don't think the fix is trivial.

**Comment by srowen** [comment]
Ah right. I was glancing at the number of commits and thought it hadn't been rebased. The problem isn't this shutdown hook but another one in `yarn.ApplicationMaster`. Dang. Well I don't think the problem is a `null` shutdown hook, and I don't think it is an error per se. Maybe it's not even worth logging because it's OK? that is, if we find `close()` is called during shutdown and so we can't remove a hook, just keep going?

**Comment by nishkamravi2** [comment]
Agreed. That's why a try-catch block seems enough for this one. Ok with removing logging.

File: core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala [hunk 1]
```diff
@@ -148,7 +148,13 @@ private[spark] class DiskBlockManager(blockManager: BlockManager, conf: SparkCon
   /** Cleanup local dirs and stop shuffle sender. */
   private[spark] def stop() {
     // Remove the shutdown hook.  It causes memory leaks if we leave it around.
-    Runtime.getRuntime.removeShutdownHook(shutdownHook)
+    if(shutdownHook != null) {
+      try {
+        Runtime.getRuntime.removeShutdownHook(shutdownHook)
+      } catch {
+        case e: Exception => None
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala]: Narrow this to `IllegalStateException`, to only squash this particular case, hopefully? I think leaving the `null` check in is OK even if I'm not sure it can happen since the variable is assigned to non-`null` in its constructor. Nit: put a space after `if`. I'll pause a beat for others to weigh in today.

File: core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala [hunk 2]
```diff
@@ -148,7 +148,13 @@ private[spark] class DiskBlockManager(blockManager: BlockManager, conf: SparkCon
   /** Cleanup local dirs and stop shuffle sender. */
   private[spark] def stop() {
     // Remove the shutdown hook.  It causes memory leaks if we leave it around.
-    Runtime.getRuntime.removeShutdownHook(shutdownHook)
+    if(shutdownHook != null) {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala]: need space after `if`

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala]: actually, when can `shutdownHook` be null?

**Comment by andrewor14** [comment]
Hey @nishkamravi2 @srowen here's another thought, what if we do
```
if (!Utils.isShutdown) {
  Runtime.getRuntime.removeShutdownHook(hook)
}
```
instead?

File: core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala (hunk 2 above)
Comment (MattWhelan) [inline core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala]: I don't think it can be.  Unless you're calling `stop` before the constructor has finished, which doesn't seem possible.

The docs for `removeShutdownHook` imply that you can pass it any thread, and it'll just return false if that thread isn't currently registered, so even double-calls are unlikely to cause problems there.  There's just the IllegalStateException, which could happen if someone else's shutdown hook is calling stop.  That's definitely worth protecting against.

**Comment by srowen** [comment]
@andrewor14 yeah I like that, although, it works by adding and removing a dummy hook and checking for an exception :) Maybe we can make a Utils method for a `safeRemoveShutdownHook` that just catches the exception? there is one other place in `ExecutorRunner` where this can be used too. I know, that starts to be more than a HOTFIX. I'd be OK either construing this as part of the original JIRA or just making a new one for this slightly different issue.

**Comment by andrewor14** [comment]
Yeah, having a `Utils.safeRemoveShutdownHook` sounds good

**Comment by nishkamravi2** [comment]
Utils.safeRemoveShutdownHook sounds ok, albeit a bit pretentious :) since a try-catch block alone doesn't quite justify creation of a utility function (unless we have a dozen or more occurrences). I would be perfectly fine with just adding a try-catch block around the other occurrence. The null-pointer check seems unnecessary (and incurs cost), so will delete it. 

Also, it seems like inShutDown function is expensive. And should be used only when necessary. I would vote for deleting all its occurrences except for the one in SparkUncaughtExceptionHandler. Unless someone feels strongly otherwise, I can do that cleanup as part of this fix.

**Comment by srowen** [comment]
I am also OK with just using the try-catch in both places it's needed, rather than make a utility function. I wouldn't remove `Utils.inShutdown`. I assume there was some reason to add it. It's expensive but only used in exception / shutdown paths. At least, I would not do this _here_.

**Comment by JoshRosen** [comment]
I still stand by my original suggestion of simply wrapping the single problematic `removeShutdownHook()` call in a `try` block: https://github.com/apache/spark/pull/4627#issuecomment-74613715.  Best to keep things simple for now and avoid introducing new bugs while trying to fix this one.

File: core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala [hunk 3]
```diff
@@ -174,11 +174,8 @@ class NewHadoopRDD[K, V](
             }
           }
         } catch {
-          case e: Exception => {
-            if (!Utils.inShutdown()) {
-              logWarning("Exception in RecordReader.close()", e)
-            }
-          }
+          case e: Exception => 
+            logWarning("Exception in RecordReader.close()", e)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala]: why this change?

**Comment by andrewor14** [comment]
@nishkamravi2 the latest changes look fine to me. I just have a minor question. We can do more refactoring outside of this hot fix and after the 1.3 release.

File: core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala (hunk 3 above)
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala]: I think we can back that out if there's any question to keep this limited. @JoshRosen Yes that's where we ended up again, now that it's clear that there are several ways and several places this can happen. It's easiest just to ignore the exception.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala]: sorry, by "back that out" do you mean revert this change or keep this change?

Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala]: Ah sorry, I mean revert. Although I kind of like the additional follow-on changes, for this PR we should foremost focus on patching up the exception. If there's any question about any extra changes, like the one you raised, I'd favor just reverting that part.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala]: ok, I agree.

**Comment by nishkamravi2** [comment]
@srowen @andrewor14  I can potentially issue a different PR for the performance issue and back it out from here (it does seem somewhat unrelated). In general, I would vote for discouraging use of this function. Code patterns like these tend to spread-- one could argue checking this condition on several other exit paths in Spark? Also, do we know that if we delete the check we would see noisy stack traces or are we just being careful speculatively?

**Comment by andrewor14** [comment]
We've had releases in the past where a simple SparkPi program would exit with a huge harmless exception that confused many users. I would prefer to keep the changes in this hot fix minimal and change the pattern separately.

Also I just realized this never ran tests, so Jenkins this is ok to test.

**Comment by srowen** [comment]
ok to test

**Comment by srowen** [comment]
This change looks like the right targeted fix. Waiting on Jenkins.

**Comment by andrewor14** [comment]
Thanks I merged this into master 1.3
